### ISSUE TITLE
Crash on cancelling "Connecting to server"

### ISSUE THREAD

**Issue by VADemon** [Pokechu22/WorldDownloader#100]
This crash occured when I pressed the cancel button on the "Connecting..." or "Loading..." screen that was stuck. The actual server is behind a Waterfall/BungeeCord lobby server.
Minecraft 1.12.2 
mod-litemodwdl-4.0.3.1-mc1.12.2.litemod
```
java.lang.NullPointerException: World Downloader Mod: Saving world
	at wdl.WDL.savePlayer(WDL.java:657)
	at wdl.WDL.saveEverything(WDL.java:574)
	at wdl.WDL.lambda$startSaveThread$2(WDL.java:437)
	at java.lang.Thread.run(Unknown Source)
```

[log-before-crash.txt](https://github.com/Pokechu22/WorldDownloader/files/2478671/log-before-crash.txt)
[crash-2018-10-15_14.08.28-client.txt](https://github.com/Pokechu22/WorldDownloader/files/2478673/crash-2018-10-15_14.08.28-client.txt)
Don't mind the presence of the other coremod, it's allowed ;)

**Comment by Pokechu22** [comment]
Cancelling on the connection screen does some [pretty weird stuff](https://bugs.mojang.com/browse/MC-74984) (read through the related issues on that ticket).  I can't say I can fix that bit of it :P

I'll look into this, though.  A brief look at that log indicates something weird already; it should have detected that you disconnected beforehand and started saving then, not when you started joining again.  Did you stop downloading before disconnecting?  Did you use the disconnect button in the pause menu before stopping?  Or something else?

**Comment by VADemon** [comment]
Luckily I have it on video: https://www.twitch.tv/videos/322618880?t=19h59m40s
It's the top right client that crashed

1. The top right client gets stuck at the same time as bottom left (network issue?)
2. Time out/kicked by the server. Auto-reconnect.
3. Auto-reconnect fails server-side and top right client is kicked with a message I haven't seen before.
4. Reconnect: Stuck in Loading... screen
5. Manual cancel, reconnect
6. WDL saving window shows and crashes the client

PS: I want to add that the automatic resume feature is essential to us, because the clients run autonomously. When the client is disconnected by server, WDL resumes automatically. But when you disconnect manually, you have to restart the download yourself.

**UPD:** Yeah maybe WDL doesn't initiate save when it's disconnected by server, but does so if client disconnects.

**Comment by VADemon** [comment]
This is what a disconnect/timeout looks like. The world is only saved upon successful reconnect because it is detected as a "world change"

So WDL doesn't detect a kick/timeout correctly to start saving?
```
[17:55:01] [main/INFO] [wdl.WDLMessages]: [WorldDL] onChunkNoLongerNeeded: Saved chunk at -2177, 3000
[17:55:01] [main/INFO] [wdl.WDLMessages]: [WorldDL] onChunkNoLongerNeeded: Saved chunk at -2177, 3001
[17:55:01] [main/INFO] [wdl.WDLMessages]: [WorldDL] onChunkNoLongerNeeded: Saved chunk at -2177, 3002
---- TIME OUT HERE ----
[17:55:08] [main/INFO] [journeymap]: Mapping halted in I:\MultiMC\instances\1.12.2-wdl_instance\.minecraft\journeymap\data\mp\2b2t\DIM0
[17:55:09] [main/INFO] [FML]: Applying holder lookups
[17:55:09] [main/INFO] [FML]: Holder lookups applied
[17:55:14] [main/INFO] [net.minecraft.client.multiplayer.GuiConnecting]: Connecting to 2b2t.org, 25565
[17:55:15] [Netty Client IO #23/INFO] [FML]: Unexpected packet during modded negotiation - assuming vanilla or keepalives : net.minecraft.network.play.server.SPacketChat
[17:55:15] [Netty Client IO #23/INFO] [FML]: Aborting client handshake "VANILLA"
[17:55:15] [Netty Client IO #23/INFO] [FML]: [Netty Client IO #23] Client side vanilla connection established
--- RECONNECTED HERE ---
[17:55:15] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] 2b2t is full
[17:55:16] [main/INFO] [net.minecraft.advancements.AdvancementList]: Loaded 15 advancements
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] World change detected. Download will start once current save completes.
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Save started.
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34712, y=34, z=47914]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34710, y=25, z=47912]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34711, y=30, z=47913]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34705, y=25, z=47980]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34710, y=19, z=47931]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34725, y=15, z=47910]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34710, y=21, z=47965]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34685, y=24, z=47927]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34681, y=15, z=47931]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34686, y=15, z=47922]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34684, y=24, z=47929]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34732, y=13, z=48001]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34729, y=13, z=48007]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34730, y=13, z=48001]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34734, y=13, z=48001]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34728, y=13, z=48010]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34697, y=21, z=47967]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Using old TE from saved file: minecraft:chest at Pos[x=-34702, y=20, z=47955]
[17:55:16] [WDL Save Thread/INFO] [wdl.WDLMessages]: [WorldDL] Old TE does not need importing: minecraft:mob_spawner at Pos[x=-34697, y=24, z=47923]
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Saving player data...
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Player data saved.
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Saving world metadata...
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] World metadata saved.
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Saving map data...
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Map data saved.
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Saving chunks...
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Chunk data saved.
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Waiting for ThreadedFileIOBase to finish...
[17:55:16] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] [WorldDL] Save complete. Starting download again.
[17:55:16] [main/INFO] [wdl.WDLMessages]: [WorldDL] Sending plugin channel registration.
[17:55:16] [main/INFO] [wdl.WDLMessages]: [WorldDL] onWorldLoad: different server!
[17:55:16] [main/INFO] [wdl.WDLMessages]: [WorldDL] Server brand = Waterfall (git:Waterfall-Bootstrap:1.12-SNAPSHOT:4330d86:184) <- Paper.  Using spigot track distances.
[17:55:17] [main/INFO] [journeymap]: Loading journeymap.topo.config
[17:55:17] [main/INFO] [journeymap]: Loaded 0 waypoints from I:\MultiMC\instances\1.12.2-wdl_instance\.minecraft\journeymap\data\mp\2b2t\waypoints
[17:55:17] [main/INFO] [journeymap]: Mapping started in I:\MultiMC\instances\1.12.2-wdl_instance\.minecraft\journeymap\data\mp\2b2t\DIM1. Memory: 645MB total, 136MB free 
[17:55:20] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] Position in queue: 80
[17:55:30] [main/INFO] [net.minecraft.client.gui.GuiNewChat]: [CHAT] Position in queue: 80
```

**Comment by Pokechu22** [comment]
That's probably the cause... automatic resume isn't actually supposed to be a thing, so it's interesting that it is automatically resuming for you.  There used to be a function several years ago to automatically start, but it was removed due to it being extremely unstable.  

The whole code with regards to detecting disconnects and dimension changes is a mess of `static` variables and has been for several years now... at some point I really need to clean all that up, but I don't think I can yet.

I can look into getting it to save on disconnect, though... I have a few ideas how I can fix that (which I thought already worked, but I guess not).

**Comment by Pokechu22** [comment]
I've fixed disconnect detection in version 4.0.3.2, with commit 8188ab5c28dcc1ca382db1accfe7002c69987bba.  This should hopefully make things a lot less buggy, but I still haven't extensively tested all of these cases.  Let me know if you run into problems (or if this makes the behavior worse for what you're doing).

**Comment by VADemon** [comment]
A lot of time passed, I noticed this is still open. I will close this, considering resolved and fixed.

---

I remember testing it a couple versions later when I found a released build and it seemed to work (fixed) on 1.12.2. However since our group **depended on exactly this bug** to seamlessly restart WDL on reconnect for automatic downloading, we remained on the old version. (I think I had pinged you on IRC about this)

I can't say for sure, but maybe the lack of explicit world saving by this mod on kick/disconnect/MC exit caused many "holes" during our WDL project. (I mean when palyers got disconnected but the mod did not detect it, did not save the world and people closed their MC client.) These void holes were missing chunks that we had to fill up retroactively (go to the void area and redo the download).

Now talking back to you, I think you should've been named along with the rest of us. You too helped us to make this happen with your mod: https://www.reddit.com/r/2b2t/comments/dzvq67/presenting_the_2b2torg_100k_mapping_project_world/

Thanks Pokechu for being an awesome, friendly and helpful lad.
